### PR TITLE
feat: add TaskStateLeave to task

### PR DESCRIPTION
**Pull request by gaius-qi**
Signed-off-by: Gaius <gaius.qi@gmail.com>

## Description
- add `TaskStateLeave` to task.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Review by yxxhero — APPROVED** [review]

File: scheduler/resource/task.go [hunk 1]
```diff
@@ -173,6 +182,10 @@ func NewTask(id, url string, taskType commonv1.TaskType, meta *commonv1.UrlMeta,
 				t.UpdateAt.Store(time.Now())
 				t.Log.Infof("task state is %s", e.FSM.Current())
 			},
+			TaskEventLeave: func(e *fsm.Event) {
+				t.UpdateAt.Store(time.Now())
+				t.Log.Infof("task state is %s", e.FSM.Current())
```
Comment (bigerous) [inline scheduler/resource/task.go:187]: Why not print more info, like:
```
t.Log.Infof("event is TaskEventLeave,  task state is %s", e.FSM.Current())
```

File: scheduler/resource/task_manager.go [hunk 2]
```diff
@@ -104,11 +104,21 @@ func (t *taskManager) RunGC() error {
...
-		if elapsed > t.ttl && task.PeerCount() == 0 && !task.FSM.Is(TaskStateRunning) {
+		if elapsed > t.ttl && task.FSM.Is(TaskStateLeave) {
 			task.Log.Info("task has been reclaimed")
 			t.Delete(task.ID)
+			return true
 		}
 
+		// If there is no peer then switch the task state to TaskStateLeave.
+		if task.PeerCount() == 0 {
+			if err := task.FSM.Event(TaskEventLeave); err != nil {
```
Comment (bigerous) [inline scheduler/resource/task_manager.go:115]: Why trigger `TaskEventLeave` here? Is it better to trigger `TaskEventLeave` when grpc of TaskLeave is called?

File: scheduler/resource/task.go (hunk 1 above)
Comment (gaius-qi) [reply scheduler/resource/task.go:187]: The log remains consistent, and I don't care what event happens when I check, only the state switch.

File: scheduler/resource/task_manager.go (hunk 2 above)
Comment (gaius-qi) [reply scheduler/resource/task_manager.go:115]: Unified logic in the GC is clear.

Comment (bigerous) [reply scheduler/resource/task_manager.go:115]: But GC is called periodically with a long interval(like 1min). It is not timely.

Comment (gaius-qi) [reply scheduler/resource/task_manager.go:115]: If it is not timely, there is no problem in letting peers go back-to-source directly. Actively triggering the seed peer is also going back-to-source.

**Review by bigerous — APPROVED** [review]
LGTM

**Pull request merged**